File: rafagas/modules/dataProcess.py
#! /usr/bin/python
import sys
import numpy as np 
import matplotlib.pyplot as plt
from offcalc import offcalc



#%%%%%%%%%5%%%%%%%%%%%%%
def dataProcess(Fs , #Sample freq
	wd = 100,
	path = "/dev/shm/",
	offman = 0,
	plotMode=False):
	ruidoR = np.fromfile(path + "ruidoR.dat",'float32')
	ruidoI = np.fromfile(path + "ruidoI.dat",'float32')
	ruidoC = ruidoR + 1j * ruidoI
	del ruidoR,ruidoI

	dataR = np.fromfile(path + "dataR.dat",'float32')
	dataI = np.fromfile(path + "dataI.dat",'float32')
	dataC = dataR + 1j * dataI
	del dataR,dataI

	#lenRRaw=len(ruidoC)
	#lenDRaw=len(dataC)





	#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
	#Calibration processing

	calibrationOffset = 0.2 *Fs  #conversion from time to samples

	# signal calibration
	offset = offcalc(dataC,Fs) + offman #received samples offset due to hardware & software lag [samples];


	dataR = np.fromfile(path + "fdataR.dat",'float32')
	dataI = np.fromfile(path + "fdataI.dat",'float32')

	dataC = dataR + 1j * dataI
	del dataR,dataI

	ruidoC = ruidoC[ int(calibrationOffset) : int(calibrationOffset*Fs) ];
	dataC  = dataC [ offset : int(offset+Fs)  ];


	lenR=len(ruidoC)
	lenD=len(dataC)


	#TODO
	#calibrated data plotting


	N=int( wd*Fs/1000000) # conversion of window duration from miliseconds to samples

	v=int(np.floor(lenD/N)) # number of pulses recorded
	PA = np.zeros((v,N)) + 1j * np.zeros((v,N))


	for i in range(0,v):# Run cross correlation for v times
	    
		x=ruidoC[int(i*N) : int(i*N+N)] #TX
		y=dataC[int(i*N)  : int(i*N+N)] # RX
		
		rxy=np.correlate( x, np.conj(y) , 'full' ) # Cross correlation of the TX and RX conjugated data
		Ryx=np.flip(rxy[0:N],0) # Flip the correlation result and take the first N samples (Ryx(t) = Rxy(-t)
		PA[i] = Ryx

	Ryx = PA.mean(axis=0)
	Ryx = np.abs(Ryx)

	if plotMode :
		plt.plot(np.abs(Ryx))
		plt.show()
	else:
		return Ryx


def main():
	Fs = 20e6 #Sample freq
	wd = 100; # window duration [us] for the correlation purpose
	path = "/dev/shm/"

	dataProcess(Fs,wd,path)


if __name__ == '__main__':
    main()

File: rafagas/matlab/dataProcess.m
% data proccessing for white noise

close all
clear all
format long

%%%%%%%%%%%%%%%%%%%%%%%


Fs=20e6; %Sample freq
wd=100; % window duration [us] for the correlation purpose
path="/dev/shm/";


%%%%%%%%%5%%%%%%%%%%%%%
fid=fopen(path+"ruidoR.dat",'rb');
ruidoR=fread(fid,'float');

fid=fopen(path+"ruidoI.dat",'rb');
ruidoI=fread(fid,'float');

ruidoC=complex(ruidoR,ruidoI);  %transmitted complex data

clear ruidoR
clear ruidoI

lenRRaw=size(ruidoC(:,1));
lenRRaw=lenRRaw(1,1);


fid=fopen(path+"dataR.dat",'rb');
dataR=fread(fid,'float');

fid=fopen(path+"dataI.dat",'rb');
dataI=fread(fid,'float');

dataC=complex(dataR,dataI); % received complex data

clear dataR
clear dataI

lenDRaw=size(dataC(:,1));
lenDRaw=lenDRaw(1,1);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%Calibration processing



calibrationOffset = 0.2 *Fs ; % conversion from time to samples


% signal calibration
offset = offsetcalc(dataC,Fs); % received samples offset due to hardware & software lag [samples]


fid=fopen(path+"fdataR.dat",'rb');
dataR=fread(fid,'float');

fid=fopen(path+"fdataI.dat",'rb');
dataI=fread(fid,'float');

dataC=complex(dataR,dataI); % filtered rx complex data

dataC  = dataC (offset+1:offset +Fs);
ruidoC = ruidoC(calibrationOffset+1:calibrationOffset+Fs);

lenR=length(ruidoC(:,1));
lenD=length(dataC(:,1));

%calibrated data plotting

figure
subplot(2,1,1)
plot(0:1/Fs:lenR/Fs-1/Fs,real(ruidoC),0:1/Fs:lenR/Fs-1/Fs,imag(ruidoC))
title('Tx calibrated')
xlabel('time [s]')
subplot(2,1,2)
plot(0:1/Fs:lenD/Fs-1/Fs,real(dataC),0:1/Fs:lenD/Fs-1/Fs,imag(dataC))
title('Rx with offset')
xlabel('time [s]')


N=wd*Fs/1000000 ; % conversion of window duration from miliseconds to samples

v=floor(lenD/N); % number of pulses recorded

y=zeros(N,v);

for i=0:v-1 % Run cross correlation for v times
    

    x=ruidoC(i*N+1:i*N+N,1); % TX
    y=dataC(i*N+1:i*N+N,1); % RX
    
    rxy=xcorr(x,conj(y)); % Cross correlation of the TX and RX conjugated data
    Ryx=flip(rxy(1:N)); % Flip the correlation result and take the first N samples (Ryx(t) = Rxy(-t)
    %hold on
    %figure
    %plot(x)
    %plot(abs(y))
    %hold off
PA(:,i+1)=Ryx'; %Store the results in an array
end
clear Ryx; 
 
Ryx=sum(PA')/v; % Average the values 



figure;
plot(0:1/Fs*1000000:N/Fs*1000000-1/Fs*1000000,abs(Ryx)); % Plot the estimated impulse response

title('Estimated Impulse Response');
xlabel('time [us]')
grid;

%%%%%%%%%%%% GUARDAR RESPUESTA AL IMPULSO %%%%%%%%%%%%%%%%%

%guardarRespAlImpulso(abs(Ryx),savePath,Fs,wd,connectionType);


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%



%[maximum,pico]=max(abs(Ryx));
%pico=pico/Fs

%f = fopen ("datos/ImpResp.dat", 'ab');
%v = fwrite (f, abs(Ryx),'float');
%fclose (f);

%&&&&&&&&&&&&&&&&&&&&&&


%modu= conv(ruidoC,Ryx);

%lenM=size(modu(:,1));
%lenM=lenM(1,1);
%figure
%plot(0:1/Fs:lenM/Fs-1/Fs,real(modu),0:1/Fs:lenM/Fs-1/Fs,imag(modu))


